File: src/code/python/data_structures/trie.py
class TrieNode:
    def __init__(self) -> None:
        self.children = {}
        self.is_word = False


class Trie:
    def __init__(self) -> None:
        self.root = TrieNode()

    def build(self, words: list[str]) -> None:
        for word in words:
            self.insert(word)

    def insert(self, word: str) -> None:
        curr = self.root

        for char in word:
            if char not in curr.children:
                curr.children[char] = TrieNode()

            curr = curr.children[char]

        curr.is_word = True

    def search(self, word: str) -> bool:
        curr = self.root

        for char in word:
            if char not in curr.children:
                return False

            curr = curr.children[char]

        return curr.is_word

    def starts_with(self, prefix: str) -> bool:
        curr = self.root

        for char in prefix:
            if char not in curr.children:
                return False

            curr = curr.children[char]

        return True

    def collect_words(self, node: TrieNode, prefix: str) -> list[str]:
        words = []

        if node.is_word:
            words.append(prefix)

        for char, child_node in node.children.items():
            words.extend(self.collect_words(child_node, prefix + char))

        return words

    def __repr__(self) -> str:
        return 'Trie:\n' + self._print_trie(self.root)

    def _print_trie(self, node: TrieNode | None, level: int = 0, prefix: str = '') -> str:
        output = ''
        prefix_str = '    ' * level + prefix

        if not node:
            return output

        if node.is_word:
            output += prefix_str + ' ├─ ' + '(*)' + '\n'

        for i, (char, child_node) in enumerate(node.children.items()):
            is_last = i == len(node.children) - 1
            marker = '└─ ' if is_last else '├─ '
            output += prefix_str + marker + char + '\n'
            output += self._print_trie(child_node, level + 1, '    │' if not is_last else '     ')

        return output

File: src/code/ruby/data_structures/trie.rb
class TrieNode
  attr_accessor :children, :is_word

  def initialize
    @children = {}
    @is_word  = false
  end
end

class Trie
  attr_accessor :root

  def initialize
    @root = TrieNode.new
  end

  def build(words)
    words.each { |word| insert(word) }
  end

  def insert(word)
    node = @root
    word.each_char do |char|
      node.children[char] ||= TrieNode.new
      node = node.children[char]
    end
    node.is_word = true
  end

  def search(word)
    node = @root
    word.each_char do |char|
      return false unless node.children.key?(char)
      node = node.children[char]
    end
    node.is_word
  end

  def starts_with(prefix)
    node = @root
    prefix.each_char do |char|
      return false unless node.children.key?(char)
      node = node.children[char]
    end
    true
  end

  def collect_words(node = @root, prefix = '')
    words = []
    words << prefix if node.is_word
    node.children.each do |char, child_node|
      words.concat(collect_words(child_node, prefix + char))
    end
    words
  end

  def to_s
    "Trie:\n" + print_trie(@root)
  end

  private

  def print_trie(node, level = 0, prefix = '')
    return '' unless node

    output = ''
    prefix_str = '    ' * level + prefix
    output += prefix_str + ' ├─ (*)' + "\n" if node.is_word

    node.children.each_with_index do |(char, child_node), i|
      is_last = i == node.children.size - 1
      marker = is_last ? '└─ ' : '├─ '
      output += prefix_str + marker + char + "\n"
      new_prefix = is_last ? '     ' : '    │'
      output += print_trie(child_node, level + 1, new_prefix)
    end

    output
  end
end
